### PR TITLE
storage: heartbeat write handles in (compute) persist_sink

### LINKED ISSUE #15191: storage: persist_sink should gracefully handle it's write handle expiring, not bring down compute
It can happen that it takes a long time (longer than the default lease timeout of 15 min) to read the initial snapshot of the collection that is being sinked. In such a case, the writer ID of the write handle that the `persist_sink` uses is being expired and any attempted writes will fail. Such failures will currently manifest as a panic! that will bring down the compute replica that the sink is running in.

Internal discussion of a situation where this problem occured: https://materializeinc.slack.com/archives/C040490DQ0N/p1664974410574429

I see multiple valid solutions:

1.  `append_batches` in `persist_sink.rs` needs to periodically heartbeat its handle.  `write_batches` in that same file already does periodically heartbeat it's handle. That operator currently only heartbeats its handle when   the frontier advances, though, so it might be subject to the same problems. It seems we need to heartbeat independent of any progress in our input.
  To make this work, we'd need to either periodically schedule the timely operator that we're hosted in or summon a tokio task that does the heartbeating. I'm currently not sure where and how that tokio task would be running when spawned inside an `async` timely operator.

2. `append_batches` needs to recognize when it's handle is expired and just create a new one. We currently create the `WriteHandle` right outside our operators main loop. Instead, we could turn that into an option and, at the start of the loop, initialize a write handle when we notice that we don't have one. We can then invalidate our write handle when we learn that we're expired, from the attempted `compare_and_append` call.

3. All persist handles can start a tokio task that periodically heartbeats until the handle is dropped. This is likely the most controversial options because it requires changing how persist handles work. We have been hesitant in the past to add this additional workload to all handles because most users of persist handles know better when they should be heartbeated.

Because of the controversial nature of option 3. I'd gravitate towards option 1. or 2. with a slight preference for 2. because it seems most resilient because we just learn to work with potential failure modes.

### PR DESCRIPTION
**Pull request by aljoscha**
Before, we were heartbeating the "writer" write handle, but only in response to new updates or the frontier advancing.

Now, we heartbeat both the "writer" and "appender" write handles and we make sure to do so independently of changes in operator inputs.

This fixes a bug that we observed in production where it can take a long time to read the initial snapshot of the sinked collection, leading to our write handle lease being expired.

I manually checked that this usage pattern of `async` timely operators does not cause us to go into a busy loop.

Fixes #15191

### Tips for reviewer

@philip-stoev I don't know if we can do much here in terms of QA. The customer env where the bug occured was restarting the compute cluster about once every 45 minutes. In general, I think it's hard to test these timeout-related bugs. Any thoughts?

cc @petrosagg Only when you're done with pg demuxing: it seems this usage pattern is ok for the async operator. We don't seem to be ending up in a busy loop.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

**Review by pH14 — COMMENTED** [review]
Write handle calls LGTM. I can't speak to the timely-async interactions though, which is the real trick here!

File: src/compute/src/sink/persist_sink.rs [hunk 1]
```diff
@@ -606,6 +609,21 @@ where
...
+                // This could lead to the "leaked blob reaper" eventually
+                // removing blobs that have been written but not yet appended to
+                // the shard.
+                //
+                write.maybe_heartbeat_writer().await;
+                trace!("persist_sink {sink_id}/{shard_id}: writer heartbeating write handle");
+                // NOTE: We schedule ourselves to make sure that we keep
+                // heartbeating even in cases where there are no changes in our
+                // inputs (updates or frontier changes).
+                activator.activate_after(Duration::from_secs(60));
```
Comment (guswynn) [inline src/compute/src/sink/persist_sink.rs:625]: does this do anything? if we have nothing to do and loop around to `scheduler.notified().await`, and we re-activate after 60 seconds, all that happens is that the `notified` future gets polled again and returns `Pending`

it seems to me that what we want is to run code every x seconds EVEN IF we aren't notified of any changes, which likely requires the new async integration, which will be hard to use here as `AsyncInputHandle` doesn't really have a good `for_each` parallel (we can probably add this)

Comment (aljoscha) [reply src/compute/src/sink/persist_sink.rs:625]: It does, yes. If you turn the logs into `info!` logging and the Duration to, say, `from_millis(10)` you will be flooded with logs. With `from_secs(60)` it's lost in the noise at steady state, with our 1sec timestamping interval.

Comment (guswynn) [reply src/compute/src/sink/persist_sink.rs:625]: interesting, i need to read the `build_async` impl more closely

**Comment by philip-stoev** [comment]
@aljoscha, this will get some testing in the Antithesis environment where network disruptions will be injected. Some day we will also have very-long-running tests with very large snapshots. But both of those things are outside of the scope of this PR.

**Comment by aljoscha** [comment]
alright, what say you? Can I get a +1 to merge from someone? :sweat_smile: Maybe @petrosagg to confirm that is is "okay" usage of the current `build_async` framework?

**Review by petrosagg — APPROVED** [review]

**Pull request merged**